src/guardian/duplicates/keys-duplication-checker.service.ts

```diff
@@ -0,0 +1,180 @@
+      if (operators.size == 1) {
+        duplicates.push(...this.findDuplicatesWithinOperator(occurrences));
+
+        continue;
+      }
+
+      // Case: Deposited keys
+      if (occurrences.some((key) => key.used)) {
+        duplicates.push(...this.filterNonDepositedKeys(occurrences));
+        continue;
+      }
+
+      // Case: Duplicates across multiple operators
+      const events = await this.fetchSigningKeyEvents(key, blockData);
+      const originalEvent = this.findOriginalEvent(events);
+
+      this.checkMissingOperators(operators, events, blockData); // Sanity check for operator events
+
+      // Case: owner of original key has duplicated keys
+      const keyOwnerKeys = occurrences.filter(
+        (key) =>
+          key.moduleAddress === originalEvent.moduleAddress &&
+          key.operatorIndex === originalEvent.operatorIndex,
+      );
+
+      const originalKey = this.findOriginalKeyWithinOperator(keyOwnerKeys);
+
+      this.logger.log('Original key is', {
+        ...{
+          originalKey,
+          createBlockNumber: originalEvent.blockNumber,
+          createBlockHash: originalEvent.blockHash,
+          createLogINdex: originalEvent.logIndex,
+        },
+        currentBlockNumber: blockData.blockNumber,
+        currentBlockhash: blockData.blockHash,
+      });
+
+      duplicates.push(
+        ...occurrences.filter((k) => !this.isSameKey(k, originalKey)),
+      );
+    }
+    return duplicates;
+  }
+
+  public findDuplicateKeys(keys: RegistryKey[]): [string, RegistryKey[]][] {
+    const keyOccurrencesMap = keys.reduce((acc, key) => {
+      const occurrences = acc.get(key.key) || [];
+      occurrences.push(key);
+      acc.set(key.key, occurrences);
+
+      return acc;
+    }, new Map<string, RegistryKey[]>());
+
+    return Array.from(keyOccurrencesMap.entries()).filter(
+      ([, occurrences]) => occurrences.length > 1,
+    );
+  }
+
+  private filterNonDepositedKeys(occurrences: RegistryKey[]): RegistryKey[] {
+    return occurrences.filter((key) => !key.used);
+  }
+
+  private findOriginalKeyWithinOperator(
+    operatorKeys: RegistryKey[],
+  ): RegistryKey {
+    return operatorKeys.reduce(
+      (prev, curr) => (prev.index < curr.index ? prev : curr),
+      operatorKeys[0],
+    );
+  }
+
+  private findDuplicatesWithinOperator(
+    operatorKeys: RegistryKey[],
+  ): RegistryKey[] {
+    // Assuming keys belong to a single operator
+    const originalKey = this.findOriginalKeyWithinOperator(operatorKeys);
+    return operatorKeys.filter((key) => key.index !== originalKey.index);
+  }
+
+  private async fetchSigningKeyEvents(
+    key: string,
+    blockData: BlockData,
+  ): Promise<SigningKeyEvent[]> {
+    const { events } =
+      await this.signingKeyEventsCacheService.getUpdatedSigningKeyEvents(
+        key,
+        blockData.blockNumber,
+        blockData.blockHash,
+      );
+    return events;
+  }
+
+  private findOriginalEvent(events: any[]): any {
```

Comment:
in this file in several places `events` are listed as `any`, let's fix it, shall we?